programs/btc_spv_api/src/spv_instruction.rs

```diff
@@ -0,0 +1,76 @@
+//! Spv proof Verification Program
+use crate::id;
+use crate::spv_state::*;
+use serde_derive::{Deserialize, Serialize};
+use solana_sdk::pubkey::Pubkey;
+use solana_sdk::instruction::{AccountMeta, Instruction};
+
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub enum SpvInstruction {
+    // Client Places request for a matching proof
+    // key 0 - Signer
+    // key 1 - Account in which to record the Request and proof
+    ClientRequest(ClientRequestInfo);
```

Comment:
why do we need a Request/Response model?  I would imagine this program only accepts valid proofs.  So any other program can verify it by just checking that its in the state.

Reply:
@aeyakovenko Request here doesn't exactly refer to a request/response model.
 ClientRequest is used by the client, which in most cases will be another contract, to post a request to recieve a proof for any given transaction, called a ProofRequest. You can think of a ProofRequest as being sort of similar to an order in an exchange,  in the sense that it's publicly visible and can be interacted with by anyone else in the ecosystem. Provers are solana clients that maintain a bitcoin node, which they use to generate and submit proofs for the current open ProofRequests to the spv program. The spv program then validates the submitted proofs and if successful, stores them to the state of the account specified by the client who requested the proof in the first place, at which point they can be accessed and verified as you described.

Reply:
@pathoam 
Got it.  I think this should work.  I was initially thinking that the ProofRequest could be of-chain and contract specific.  But this might be easier for developers to access.

Reply:
@aeyakovenko 
That's the idea, I want verifying a transaction to be as simple as making a call to the spv program with the txhash. That way users and application developers don't have to worry about the burden of actually generating the proofs themselves since it's a pretty out of band process from the rest of the stack.

Reply:
How does a smart contract learn about the txhash? Typically consumers are interested in events meeting certain characteristics (spends a specific outpoint, or pays a specific person), rather than a specific known transaction

Reply:
it depends on the application, you might have to actively submit it after originating the bitcoin transaction in question in some cases. This is only a temporary measure though, the plan is to offer the option to submit ProofRequests with bloom filters as the identifying factor that way you can submit an order to verify say, all the transactions originating from a given address. Should also add an element of privacy in the sense that you could pay a bit more in fees to provers to include a more broad filter that doesn't give away direct correlation between your accounts.
That adds some more complexity to the request verification though, so I'd rather wait until we have the base product further along to get into that.

Reply:
the most useful thing you can do is commit to an acceptance predicate. Bloom filters are not useful in practice, and certainly won't preserve privacy in an on-chain setting. Typically a cross-chain protocol is not looking to mirror Bitcoin state to a contract (except may the header chain). It's looking to take action based on Bitcoin events

Reply:
I don't mean preserving privacy in the sense of making the requests opaque to outside scrutiny, since it's on chain and relies on proof requests being filled by 3rd parties. I do mean though that it offers the option to muddy the waters against cross chain account correlation. That being said, that relies on assumptions around the rest of the contract stack involved in a given call to the point that it might not be meaningful in practice. 
And to clarify, we're not talking about mirroring bitcoin chain state beyond blockheaders. The idea of using bloom filters is just intended to mimic the standard bip 37 format used by spv clients  to specify transaction(s) as opposed to having separate logic paths and representation for each selection modality. The process will be that clients advertise to receive verification of events matching their filter, at which point the provers are responsible for applying that filter and submitting matching proofs to the spv program, which validates them and if successful saves them to the data of the account submitted by the client for that purpose. The end result is that only the specifically requested events and proof of their verification end up on chain and client contracts can monitor the status of their requests by checking the state of their "submission account".

Reply:
> The idea of using bloom filters is just intended to mimic the standard bip 37 format used by spv clients to specify transaction(s) as opposed to having separate logic paths and representation for each selection modality.

Note that bip37 is disabled by default going forward (see [this PR](https://github.com/bitcoin/bitcoin/pull/16152))